pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// ConcurrencyStateHandler tracks the in flight requests for the pod. When the requests
+// drop to zero, it runs the `pause` function, and when requests scale up from zero, it
+// runs the `resume` function. If either of `pause` or `resume` are not passed, it runs
+// the respective local function(s). The local functions are the expected behavior; the
+// function parameters are enabled primarily for testing purposes.
+func ConcurrencyStateHandler(h http.Handler, logger *zap.SugaredLogger, pause, resume func(string) error, endpoint string) http.HandlerFunc {
+	logger.Info("Concurrency state endpoint set, tracking request counts, using endpoint ", endpoint)
+
+	type req struct {
+		w http.ResponseWriter
+		r *http.Request
+
+		done chan struct{}
+	}
+
+	reqCh := make(chan req)
+	doneCh := make(chan struct{})
+	go func() {
+		inFlight := 0
+
+		// This loop is entirely synchronous, so there's no cleverness needed in
+		// ensuring open and close dont run at the same time etc. Only the
+		// delegated ServeHTTP is done in a goroutine.
+		for {
+			select {
+			case <-doneCh:
+				inFlight--
+				if inFlight == 0 {
+					logger.Info("Requests dropped to zero")
+					err := pause(endpoint)
+					if err != nil {
+						logger.Errorf("Issue handling pause request: %s", err)
```

Comment:
Im not convinced this is better than the panic tbh. We need to think about how we actually want to behave here. Should we retry the request? Should we exit the entire server? Panic may be all we can do, but I think we should decide explicitly if that's the case. This might be something to talk through on Wednesday / in an ad-hoc meeting actually

Reply:
Agreed. I sketched out some thoughts [here](https://github.com/knative/serving/pull/11802#discussion_r691355399). This was a start along those lines, but didn't want to go too far before we discussed.

Reply:
oops missed that comment, let me go read now!